py4DSTEM/process/classification/featurization.py

```diff
@@ -0,0 +1,637 @@
+        iters (int): number of iterations
+        random_state (int): random state
+        return_all (bool): Whether or not to return all of the iterations in 'iters' - default is to return
+        all outputs for consensus clustering
+        """
+        Ws, Hs, W, H = {}, {}, {}, {}
+        for i in range(len(keys)):
+            self.Ws[keys[i]], self.Hs[keys[i]], self.W[keys[i]], self.H[keys[i]] = _nmf_single(
+                self.features[keys[i]],
+                max_components=max_components[keys[i]],
+                merge_thresh = merge_thresh[keys[i]],
+                iters = iters[keys[i]], random_state = random_state)
+        return
+        
+    def GMM(self, keys, cv, components, iters, random_state = None):
+        """
+        Performs gaussian mixture model on input features
+        
+        Args:
+        keys (list of str): List of strings associated with features
+        cv (str): covariance type - must be 'spherical', 'tied', 'diag', or 'full'
+        components (int): number of components
+        iters (int): number of iterations
+        random_state (int): random state
+        """
+        gmm, gmm_labels, gmm_proba = {}, {}, {}
+        for i in range(len(keys)):
+            self.gmm[keys[i]], self.gmm_labels[keys[i]], self.gmm_proba[keys[i]] = _gmm_single(self.features[keys[i]],
+                                                                                cv=cv[keys[i]],
+                                                                                components = components[keys[i]],
+                                                                                iters = iters[keys[i]], 
+                                                                                random_state = random_state)
+        return
+
+    def get_class_DPs(self, keys, classification_method, thresh, dc):
+        """
+        Returns weighted class patterns based on classification instance
+        dc (datacube) must be vectorized in real space (shape = (R_Nx * R_Ny, 1, Q_Nx, Q_Ny)
+        
+        Args:
+        keys (list of str): List of strings associated with features
+        classification_method (str): either 'nmf' or 'gmm' - finds location of clusters
+        dc (datacube) Vectorized in real space, with shape (R_Nx * R_Ny, 1, Q_Nx, Q_Ny)
+        """
+        for i in range(len(keys)):
+            class_patterns = []
+            if classification_method[keys[i]] == 'nmf':
+                for l in range(self.W[keys[i]].shape[1]):
+                    class_pattern = np.zeros((dc.data.shape[2], dc.data.shape[3]))
+                    x_ = np.where(self.W[keys[i]][:,l] > thresh[keys[i]])[0]
+                    for x in range(x_.shape[0]):
+                        class_pattern += dc.data[x_[x],0] * self.W[keys[i]][x_[x],l]
+                    class_patterns.append((class_pattern - np.min(class_pattern)) 
+                                          / (np.max(class_pattern) - np.min(class_pattern)))
```

Comment:
i'd suggest changing the normalization to divide by the sum of the weights `W[...]`